v5/core/parameterized_url.go

```diff
@@ -0,0 +1,83 @@
+package core
+
+// (C) Copyright IBM Corp. 2021.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+//
+// constructServiceURL returns a service URL that is constructed by formatting a parameterized URL.
+//
+// Parameters:
+//
+// parameterizedUrl: URL that contains variable placeholders, e.g. "{scheme}://ibm.com".
+//
+// defaultUrlVariables: map from variable names to default values.
+//		Each variable in the parameterized URL must have a default value specified in this map.
+//
+// providedUrlVariables: map from variable names to desired values.
+// 		If a variable is not provided in this map,
+// 		the default variable value will be used instead.
+//
+func constructServiceURL(
+	parameterizedUrl string,
+	defaultUrlVariables map[string]string,
+	providedUrlVariables map[string]string,
+) (string, error) {
+
+	// If null was passed, we set the variables to an empty map.
+	// This results in all default variable values being used.
+	if providedUrlVariables == nil {
```

Comment:
Incidentally enough, it's ok to read from a nil map.  You just get back the zero value ("" for string) for any key that you try to retrieve.
Reference: https://blog.golang.org/maps

It's ok to account for a nil map the way you have done here; just thought I'd mention it.